File: src/editor.c
#include "include/editor.h"
#include "include/ascii.h"
#include "include/buffer.h"
#include "include/commander.h"
#include "include/cursor.h"
#include "include/debug.h"
#include "include/display.h"
#include "include/goto.h"
#include "include/str.h"
#include "include/timeline.h"
#include "include/tui.h"
#include <curses.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void
editor__render_tui() {
    tui__render_editor();
    tui__render_statusbar();
    tui__refresh_displays();
}

editor_t
editor__new(buffer_t buffer) {
    return (editor_t) {
        .exit = false,
        .mode = MODE__NORMAL,
        .display = EDITOR_DISPLAY,
        .finder = finder__new(),
        .buffer = buffer,
        .display_list = display_list__new(),
        .commander = commander__new(),
        .timeline = timeline__new(),
    };
}

void
editor__run() {
    tui__start();

    while (!EDITOR.exit) {
        int ch = getch();

        editor__compute(ch);
        editor__render_tui();
    }

    editor__release();

    tui__exit();
}

void
editor__compute(const int ch) {
    switch (EDITOR.mode) {
        case MODE__GOTO:
            switch (ch) {
                case 'g':
                    goto__start_file();
                    break;
                case 'G':
                    goto__end_file();
                    break;
            }
            editor__set_mode(MODE__NORMAL);
            break;
        case MODE__NORMAL:
        case MODE__VISUAL:
            switch (ch) {
                case 'i':
                    editor__set_mode(MODE__INSERT);
                    break;
                case 'v':
                    editor__set_mode(MODE__VISUAL);
                    break;
                case ':':
                    editor__set_mode(MODE__COMMAND);
                    break;
                case 'j':
                case KEY_DOWN:
                    editor__mv_cursor(MOVEMENT__DOWN);
                    break;
                case 'k':
                case KEY_UP:
                    editor__mv_cursor(MOVEMENT__UP);
                    break;
                case 'h':
                case KEY_LEFT:
                    editor__mv_cursor(MOVEMENT__LEFT);
                    break;
                case 'l':
                case KEY_RIGHT:
                    editor__mv_cursor(MOVEMENT__RIGHT);
                    break;
                case 'g':
                    editor__set_mode(MODE__GOTO);
                    break;
                case '$':
                    goto__end_line();
                    break;
                case '0':
                    goto__start_line();
                    break;
                case ASCII_ESCAPE:
                    editor__set_mode(MODE__NORMAL);
                    break;
            }
            break;
        case MODE__INSERT:
            switch (ch) {
                case ASCII_ESCAPE:
                    editor__set_mode(MODE__NORMAL);
                    break;
                case KEY_BACKSPACE:
                    editor__delete_char();
                    break;
                case ENTER: {
                    editor__newline();
                    cursor_t *cursor = editor__get_cursor();
                    cursor__from(0, cursor->y + 1);
                } break;
                case KEY_UP:
                    editor__mv_cursor(MOVEMENT__UP);
                    break;
                case KEY_RIGHT:
                    editor__mv_cursor(MOVEMENT__RIGHT);
                    break;
                case KEY_DOWN:
                    editor__mv_cursor(MOVEMENT__DOWN);
                    break;
                case KEY_LEFT:
                    editor__mv_cursor(MOVEMENT__LEFT);
                    break;
                default:
                    editor__add_char(ch);
                    break;
            }
            break;
        case MODE__COMMAND:
            switch (ch) {
                case ENTER:
                    commander__eval();
                    break;
                case KEY_BACKSPACE:
                    string__remove(&EDITOR.commander.cmd, EDITOR.commander.cmd.len);
                    break;
                case ASCII_ESCAPE:
                    editor__set_mode(MODE__NORMAL);
                    break;
                default:
                    string__append(&EDITOR.commander.cmd, EDITOR.commander.cmd.len, ch);
                    break;
            }
            break;
    }
}

void
editor__mv_cursor(movement_t movement) {
    switch (movement) {
        case MOVEMENT__UP:
            cursor__up();
            break;
        case MOVEMENT__DOWN:
            cursor__down();
            break;
        case MOVEMENT__RIGHT:
            cursor__right();
            break;
        case MOVEMENT__LEFT:
            cursor__left();
            break;
    }
}

void
editor__set_mode(editor_mode_t mode) {
    EDITOR.mode = mode;
}

cursor_t *
editor__get_cursor() {
    return &EDITOR.display_list.data[EDITOR.display].cursor;
}

void
editor__newline() {
    EDITOR.buffer.data = EDITOR.buffer.data == NULL
                             ? calloc(1, sizeof(str_t))
                             : realloc(EDITOR.buffer.data, sizeof(str_t) * (EDITOR.buffer.lines + 1));

    str_t line = string__new(1);
    EDITOR.buffer.data[EDITOR.buffer.lines] = line;
    EDITOR.buffer.lines += 1;
}

str_t *
editor__get_current_line() {
    cursor_t *cursor = editor__get_cursor();

    if (cursor->y < EDITOR.buffer.lines) {
        return &EDITOR.buffer.data[cursor->y];
    }

    return NULL;
}

void
editor__add_char(int ch) {
    if (editor__get_current_line() == NULL) {
        editor__newline();
    }

    cursor_t *cursor = editor__get_cursor();
    str_t *line = editor__get_current_line();
    string__append(line, cursor->x, (char) ch);

    editor__mv_cursor(MOVEMENT__RIGHT);
}

void
editor__delete_char() {
    cursor_t *cursor = editor__get_cursor();
    str_t *line = editor__get_current_line();

    if (line != NULL) {
        string__remove(line, cursor->x);
    }

    editor__mv_cursor(MOVEMENT__LEFT);
}

const char *
editor__mode_as_str(editor_mode_t mode) {
    const char *modes[] = {
        [MODE__INSERT] = "INSERT",
        [MODE__VISUAL] = "VISUAL",
        [MODE__NORMAL] = "NORMAL",
        [MODE__COMMAND] = "COMMAND",
        [MODE__GOTO] = "GOTO",
    };

    return modes[mode];
}

void
editor__release() {
    finder__release();
    commander__release();
    timeline__release();
    display_list__release();
}

File: src/editor.go
package main

import (
	"strings"

	"github.com/gdamore/tcell/v2"
)

type Mode int

const (
	Mode__Normal = iota
	Mode__Insert
	Mode__Visual
	Mode__Command
	Mode__Goto
)

type Movement int

const (
	Movement__Up = iota
	Movement__Right
	Movement__Down
	Movement__Left
)

type LineMode int

const (
	LineMode__Normal = iota
	LineMode__Relative
)

type Config struct {
	lineMode  LineMode
	useArrows bool
}

type Editor struct {
	exit     bool
	mode     Mode
	cmdline  []rune
	finder   Finder
	buffer   Buffer
	timeline Timeline
	tui      TUI
	config   Config
}

func NewEditor() Editor {
	return Editor{
		exit:     false,
		mode:     Mode__Normal,
		cmdline:  make([]rune, 0),
		finder:   NewFinder(),
		buffer:   NewBuffer(),
		timeline: NewTimeline(),
	}
}

func (self *Editor) Run() {
	S := self.tui.screen
	defer S.Fini()

	go func() {
		for {
			Events <- S.PollEvent()
		}
	}()

	for !self.exit {
		event := <-Events

		self.compute(event)
		self.tui.Rehydrate()
	}
}

func (self *Editor) AttachIface(tui TUI) {
	self.tui = tui
}

func (self *Editor) SetConfig(config Config) {
	self.config = config
}

func (self *Editor) GetModeAsStr() string {
	modes := map[Mode]string{
		Mode__Normal:  "NORMAL",
		Mode__Insert:  "INSERT",
		Mode__Visual:  "VISUAL",
		Mode__Command: "COMMAND",
		Mode__Goto:    "GOTO",
	}

	return modes[self.mode]
}

func (self *Editor) GetCursor() *Cursor {
	return &self.buffer.cursor
}

func (self *Editor) GetCurrentLine() []rune {
	cursor := self.GetCursor()

	if self.buffer.lines <= cursor.y {
		self.addNewLine()
	}

	return self.buffer.data[cursor.y]
}

func (self *Editor) GetNlineByMode(mode int) {

}

func (self *Editor) compute(event tcell.Event) {
	S := self.tui.screen

	switch event := event.(type) {
	case *tcell.EventResize:
		S.Sync()
		break
	case *tcell.EventMouse:
		switch event.Buttons() {
		case tcell.WheelUp:
			self.exit = true
			// self.scrollUp()
			break
		case tcell.WheelDown:
			// self.scrollDown()
			break
		}
		break
	case *tcell.EventKey:
		switch event.Key() {
		case tcell.KeyUp:
			if self.config.useArrows {
				self.mvCursor(Movement__Up)
			}
			break
		case tcell.KeyRight:
			if self.config.useArrows {
				self.mvCursor(Movement__Right)
			}
			break
		case tcell.KeyDown:
			if self.config.useArrows {
				self.mvCursor(Movement__Down)
			}
			break
		case tcell.KeyLeft:
			if self.config.useArrows {
				self.mvCursor(Movement__Left)
			}
			break
		case tcell.KeyEscape:
			if self.mode != Mode__Normal {
				self.setMode(Mode__Normal)
			} else {
				self.exit = true
			}
			break
		case tcell.KeyEnter:
			switch self.mode {
			case Mode__Insert:
				cursor := self.GetCursor()

				cursor.From(0, cursor.y+1)
				break
			case Mode__Command:
				self.exeCmd()
				break
			}
			break
		case tcell.KeyBackspace, tcell.KeyBackspace2:
			if self.mode == Mode__Insert {
				self.rmChar()
			} else if self.mode == Mode__Command {
                self.cmdline = self.cmdline[:len(self.cmdline) - 1]
			}
			break
		case tcell.KeyRune:
			ch := event.Rune()

			switch self.mode {
			case Mode__Goto:
				switch ch {
				case 'g':
					GotoStartFile()
					break
				case 'G':
					GotoEndFile()
					break
				}
				self.setMode(Mode__Normal)
				break
			case Mode__Normal, Mode__Visual:
				switch ch {
				case 'i':
					EDITOR.setMode(Mode__Insert)
					break
				case 'v':
					EDITOR.setMode(Mode__Visual)
					break
				case ':':
					self.cmdline = append(self.cmdline, ':')
					EDITOR.setMode(Mode__Command)
					break
				case 'g':
					EDITOR.setMode(Mode__Goto)
					break
				case 'k':
					EDITOR.mvCursor(Movement__Up)
					break
				case 'l':
					EDITOR.mvCursor(Movement__Right)
					break
				case 'j':
					EDITOR.mvCursor(Movement__Down)
					break
				case 'h':
					EDITOR.mvCursor(Movement__Left)
					break
				case '$':
					GotoEndLine()
					break
				case '0':
					GotoStartLine()
					break
				}
				break
			case Mode__Insert:
				self.addChar(ch)
				break
			case Mode__Command:
				self.cmdline = append(self.cmdline, ch)
				break
			}
		}
	}
}

func (self *Editor) mvCursor(move Movement) {
	cursor := self.GetCursor()

	switch move {
	case Movement__Up:
		cursor.Up()
		break
	case Movement__Right:
		cursor.Right()
		break
	case Movement__Down:
		cursor.Down()
		break
	case Movement__Left:
		cursor.Left()
		break
	}
}

func (self *Editor) setMode(mode Mode) {
	self.mode = mode
}

func (self *Editor) addNewLine() {
	self.buffer.data = append(self.buffer.data, make([]rune, 0))
	self.buffer.lines += 1
}

func (self *Editor) addChar(ch rune) {
	cursor := self.GetCursor()
	line := self.GetCurrentLine()

	line = append(line, ch)
	self.buffer.data[cursor.y] = line

	self.buffer.dirty = true

	self.mvCursor(Movement__Right)
}

// FIXME
func (self *Editor) rmChar() {
	cursor := self.GetCursor()
	line := self.GetCurrentLine()

	line = append(line[cursor.x-1:], line[:cursor.x]...)
}

func (self *Editor) GetNlineOffset() int64 {
	nline := self.buffer.lines

	if nline == 0 {
		return 1
	}

	x := int64(0)

	for nline != 0 {
		nline /= 10
		x += 1
	}

	return x
}

func (self *Editor) exeCmd() {
	cmd := string(self.cmdline[1:])

	if strings.Compare("q", cmd) == 0 || strings.Compare("Q", cmd) == 0 {
		self.exit = true
	}

	self.cmdline = self.cmdline[:0]
	self.setMode(Mode__Normal)
}

func (self *Editor) checkByDirtyBuffer() {
	for i := &self.buffer; i.next != nil; i = i.next {
		// TODO
	}
}
